### PR TITLE
Make module trace emission more robust.

### PR DESCRIPTION
**Pull request by varungandhi-apple**
See individual commit messages for details.

Fixes rdar://problem/56073580.
Fixes rdar://problem/55752426.

Jordan shares most of the context here, so it is probably easiest for him to review the patch, but tagging other people to increase the bus number for trace related info (or ask questions in case something is unclear).

cc @adrian-prantl to make sure the test change for DebugInfo is ok.

### CODE REVIEW

**Comment by varungandhi-apple** [comment]
@swift-ci please smoke test

**Review by beccadax — COMMENTED** [review]
Just cosmetic things.

File: lib/Serialization/SerializedModuleLoader.cpp [hunk 1]
```diff
@@ -344,6 +344,7 @@ std::error_code SerializedModuleLoaderBase::openModuleFiles(
 std::error_code SerializedModuleLoader::findModuleFilesInDirectory(
     AccessPathElem ModuleID, StringRef DirPath, StringRef ModuleFilename,
     StringRef ModuleDocFilename, StringRef ModuleSourceInfoFileName,
+    SmallVectorImpl<char> *__ModuleInterfacePath,
```
[the remark this answers is not in the document]
Comment (varungandhi-apple) [reply lib/Serialization/SerializedModuleLoader.cpp]: Because the argument is ignored. I didn't see a convention for it anywhere else, so I used the Python convention.

File: lib/Serialization/SerializedModuleLoader.cpp [hunk 2]
```diff
@@ -818,9 +823,14 @@ void swift::serialization::diagnoseSerializedASTLoadFailure(
 bool SerializedModuleLoaderBase::canImportModule(
     std::pair<Identifier, SourceLoc> mID) {
   // Look on disk.
+  SmallVector<char, 0> *moduleInterfacePath = nullptr;
+  std::unique_ptr<llvm::MemoryBuffer> *moduleBuffer = nullptr;
+  std::unique_ptr<llvm::MemoryBuffer> *moduleDocBuffer = nullptr;
+  std::unique_ptr<llvm::MemoryBuffer> *moduleSourceInfoBuffer = nullptr;
```
[the remark this answers is not in the document]
Comment (varungandhi-apple) [reply lib/Serialization/SerializedModuleLoader.cpp]: (1) It felt a bit too noisy with all the comments (lots of `/* */`) interleaved with the 7 or 8 arguments, so I put them up earlier.
(2) It leads to poorer readability thanks to uniform syntax highlighting for comments.
(3) In case the arguments get swapped or something, we won't have an implicit conversion this way. Some of them are still the same types, so it doesn't help there 🤷‍♂.

File: lib/Serialization/SerializedModuleLoader.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (brentdax) [reply lib/Serialization/SerializedModuleLoader.cpp]: The C++ standard [reserves `__` identifiers for the compiler's use](https://timsong-cpp.github.io/cppwp/n4659/lex.name#3.1), so we probably shouldn't use it here. In practice, I think there's no convention for this because nobody particularly sees a need to indicate it in our code bases.

**Comment by varungandhi-apple** [comment]
@swift-ci please smoke test

File: lib/FrontendTool/FrontendTool.cpp [hunk 3]
```diff
@@ -237,12 +237,151 @@ template <> struct ObjectTraits<LoadedModuleTraceFormat> {
...
+
+      // If we have a swiftmodule next to an interface, that interface path will
+      // be saved (not checked), so don't save the path to this swiftmodule.
+      SmallString<256> moduleAdjacentInterfacePath(depPath);
+      computeAdjacentInterfacePath(moduleAdjacentInterfacePath);
+      if (pathToModuleDecl.find(moduleAdjacentInterfacePath)
+          != pathToModuleDecl.end())
+        continue;
+
+      if (!fs::real_path(moduleAdjacentInterfacePath, buffer,
```
Comment (adrian-prantl) [inline lib/FrontendTool/FrontendTool.cpp]: The documentation of real_path doesn't really say that it clears buffer.. perhaps use a different local variable here or explicitly clear it?

Comment (varungandhi-apple) [reply lib/FrontendTool/FrontendTool.cpp]: In a way, it doesn't really matter, because we aren't reading the buffer back. But yeah, if we were to use it, say for debugging or something, then having the code correct would be good.

Comment (varungandhi-apple) [reply lib/FrontendTool/FrontendTool.cpp]: I take that back. We are reusing the buffer across loop iterations, so we should clear it after use.

**Comment by varungandhi-apple** [comment]
@swift-ci please smoke test

File: lib/FrontendTool/FrontendTool.cpp [hunk 4]
```diff
@@ -237,12 +237,154 @@ template <> struct ObjectTraits<LoadedModuleTraceFormat> {
...
 
+namespace {
+
+// The most interesting/tricky thing here is _which_ paths get recorded. It
+// depends on how the module was synthesized. The key points are:
+//
+// 1. Paths in the cache are not recorded - Precondition: the corresponding
+//    path to the swiftinterface must already be present as a key in
+//    pathToModuleDecl.
+// 2. swiftmodules next to a swiftinterface are saved if they are up-to-date.
```
Comment (jrose-apple) [inline lib/FrontendTool/FrontendTool.cpp]: This comment is missing a bunch of context for a reader just coming upon it. What paths are you talking about? What cache?

File: include/swift/Frontend/ModuleInterfaceLoader.h [hunk 5]
```diff
@@ -150,6 +150,7 @@ class ModuleInterfaceLoader : public SerializedModuleLoaderBase {
     AccessPathElem ModuleID, StringRef DirPath, StringRef ModuleFilename,
     StringRef ModuleDocFilename,
     StringRef ModuleSourceInfoFilename,
+    SmallVectorImpl<char> *ModuleInterfacePath,
```
Comment (jrose-apple) [inline include/swift/Frontend/ModuleInterfaceLoader.h]: Yuck, but okay, I see that this is probably the only way to feed it through. Maybe we should make it more generic though: "FoundPath" or "PathForDependencyTracking" or something. What do you think?

File: lib/FrontendTool/FrontendTool.cpp [hunk 6]
```diff
@@ -237,12 +237,154 @@ template <> struct ObjectTraits<LoadedModuleTraceFormat> {
...
+    auto moduleFileType =
+      file_types::lookupTypeForExtension(path::extension(depPath));
+    auto isSwiftmodule =
+      moduleFileType == file_types::TY_SwiftModuleFile;
+    auto isSwiftinterface =
+      moduleFileType == file_types::TY_SwiftModuleInterfaceFile;
+
+    if (isSwiftmodule || isSwiftinterface) {
+      auto dep = pathToModuleDecl.find(depPath);
+      if (dep != pathToModuleDecl.end()) {
```
Comment (jrose-apple) [inline lib/FrontendTool/FrontendTool.cpp]: Can you use early exits more? Both this and the containing `if` seem like they can get the error handling out of the way and `continue`.

File: lib/Serialization/SerializedModuleLoader.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (jrose-apple) [reply lib/Serialization/SerializedModuleLoader.cpp]: Maybe it could be clearer if you named the locals in a way that communicates the intent a little more? `unusedBuffer`, `ignoreDocBuffer`, whatever.

File: lib/Serialization/SerializedModuleLoader.cpp [hunk 7]
```diff
@@ -865,8 +876,13 @@ SerializedModuleLoaderBase::loadModule(SourceLoc importLoc,
   Ctx.LoadedModules[moduleID.first] = M;
   SWIFT_DEFER { M->setHasResolvedImports(); };
 
-  if (!loadAST(*M, moduleID.second, std::move(moduleInputBuffer),
-               std::move(moduleDocInputBuffer), std::move(moduleSourceInfoInputBuffer),
+  StringRef moduleInterfacePathStr =
+    Ctx.AllocateCopy(StringRef(moduleInterfacePath.c_str(),
+                               moduleInterfacePath.size()));
```
Comment (jrose-apple) [inline lib/Serialization/SerializedModuleLoader.cpp]: There's an AllocateCopy overload for StringRefs already. (Also, good catch that you need to copy the string into long-term storage.)

**Review by jrose-apple — COMMENTED** [review]
Some initial comments, though I want to take a closer look at `computeSwiftModuleTraceInfo` still.

File: include/swift/Frontend/ModuleInterfaceLoader.h (hunk 5 above)
Comment (varungandhi-apple) [reply include/swift/Frontend/ModuleInterfaceLoader.h]: That sounds reasonable. I could rename it to `FoundModuleInterfacePath`.

Comment (varungandhi-apple) [reply include/swift/Frontend/ModuleInterfaceLoader.h]: Oh, I see what you mean. I don't see why making it more generic is useful, because then it becomes less clear what the semantics of that path are. If let's say in the future, we want a bunch of other paths out, maybe we can create a struct? (Actually, I was debating creating a struct to contain all the outparams...)

File: lib/FrontendTool/FrontendTool.cpp (hunk 6 above)
Comment (varungandhi-apple) [reply lib/FrontendTool/FrontendTool.cpp]: I can get rid of the outer one but this one seems necessary, because in both cases, we have some complicated logic. For the outer-one, there was nothing to be done, if the condition was false, so we can simply `continue`.

Comment (varungandhi-apple) [reply lib/FrontendTool/FrontendTool.cpp]: I see what you're saying. Instead of 

```
if (good) {
  happy()
  continue;
}
handleError();
```

I should change it to

```
if (!good) {
  handleError();
  continue;
}

happy();
```

File: lib/Serialization/SerializedModuleLoader.cpp (hunk 7 above)
Comment (varungandhi-apple) [reply lib/Serialization/SerializedModuleLoader.cpp]: I don't follow... are you asking me to get rid of the explicit `StringRef` constructor? If I do that, how will it call the correct `AllocateCopy` overload?

File: lib/FrontendTool/FrontendTool.cpp (hunk 4 above)
Comment (varungandhi-apple) [reply lib/FrontendTool/FrontendTool.cpp]: Good catch. How about I rephrase it to

> Paths pointing to swiftmodules on the module cache path or in the prebuilt cache are not recorded

Does that sound reasonable? More context needed?

File: include/swift/Frontend/ModuleInterfaceLoader.h (hunk 5 above)
Comment (jrose-apple) [reply include/swift/Frontend/ModuleInterfaceLoader.h]: I was thinking it's "overriding file name for debugging, dependency tracking, and other purposes that want to ask for a single path that represents this module", but that's a little long. 😄

File: lib/FrontendTool/FrontendTool.cpp (hunk 6 above)
Comment (jrose-apple) [reply lib/FrontendTool/FrontendTool.cpp]: Either way is fine if they're both complicated, I gues. I thought the error path was going to be shorter here but maybe I'm misreading it.

File: lib/Serialization/SerializedModuleLoader.cpp (hunk 7 above)
Comment (jrose-apple) [reply lib/Serialization/SerializedModuleLoader.cpp]: Ah, right, `moduleInterfacePath` is a SmallString, which is also a SmallVectorImpl. Still, `moduleInterfacePath.str()` would be simpler.

**Comment by varungandhi-apple** [comment]
@swift-ci please smoke test

File: lib/FrontendTool/FrontendTool.cpp (hunk 4 above)
Comment (jrose-apple) [reply lib/FrontendTool/FrontendTool.cpp]: I've never heard "on" used to refer to directories, only filesystems/mounts/disks.

Comment (varungandhi-apple) [reply lib/FrontendTool/FrontendTool.cpp]: Changed it to "Paths to swiftmodules in the module cache or in the prebuilt cache"

Comment (jrose-apple) [reply lib/FrontendTool/FrontendTool.cpp]: I still think this comment doesn't make sense for someone just reading along. To put it another way, if you made this a proper doc comment, it's missing the very first part that describes what the function does.

File: lib/FrontendTool/FrontendTool.cpp [hunk 8]
```diff
@@ -237,12 +237,156 @@ template <> struct ObjectTraits<LoadedModuleTraceFormat> {
...
+
+  std::string errMsg;
+  llvm::raw_string_ostream err(errMsg);
+
+  auto errorUnexpectedPath = [&err, &pathToModuleDecl]() {
+    err << "The module <-> path mapping we have is:\n";
+    for (auto &m: pathToModuleDecl)
+      err << m.second->getName() << " <-> " << m.first << '\n';
+    llvm::report_fatal_error(err.str());
+  };
```
Comment (jrose-apple) [inline lib/FrontendTool/FrontendTool.cpp]: I really don't like this "print more stuff to the error string and then error" idiom, doubly so when it's relying on capturing the error string stream. Can you make a PrettyStackTrace subclass for dumping this mapping instead? (This can happen in a follow-up PR though.)

File: lib/FrontendTool/FrontendTool.cpp [hunk 9]
```diff
@@ -237,12 +237,156 @@ template <> struct ObjectTraits<LoadedModuleTraceFormat> {
...
+    // built a swiftmodule from that interface, so we should have that
+    // filename available.
+    if (isSwiftinterface) {
+      err << "Unexpected path for swiftinterface file:\n" << depPath << "\n";
+      errorUnexpectedPath();
+    }
+
+    // Skip cached modules in the prebuilt cache. We will add the corresponding
+    // swiftinterface from the SDK directly, but this isn't checked. :-/
+    if (StringRef(depPath).contains(prebuiltCachePath))
```
Comment (jrose-apple) [inline lib/FrontendTool/FrontendTool.cpp]: `startswith` is probably good enough.

File: lib/FrontendTool/FrontendTool.cpp [hunk 10]
```diff
@@ -237,12 +237,156 @@ template <> struct ObjectTraits<LoadedModuleTraceFormat> {
...
+      continue;
+
+    // If we have a swiftmodule next to an interface, that interface path will
+    // be saved (not checked), so don't save the path to this swiftmodule.
+    SmallString<256> moduleAdjacentInterfacePath(depPath);
+    computeAdjacentInterfacePath(moduleAdjacentInterfacePath);
+    if (pathToModuleDecl.find(moduleAdjacentInterfacePath)
```
Comment (jrose-apple) [inline lib/FrontendTool/FrontendTool.cpp]: You can get a slight simplification here with `count` instead of `find`.

File: lib/FrontendTool/FrontendTool.cpp [hunk 11]
```diff
@@ -237,12 +237,156 @@ template <> struct ObjectTraits<LoadedModuleTraceFormat> {
...
+        != pathToModuleDecl.end())
+      continue;
+
+    if (!fs::real_path(moduleAdjacentInterfacePath, buffer,
+                       /*expand_tilde*/true)) {
```
Comment (jrose-apple) [inline lib/FrontendTool/FrontendTool.cpp]: This isn't correct because it's not going through the VFS. I'm not sure what it's checking anyway; `fs::real_path` isn't the usual way to see if something exists (that'd be `fs::exists`).

Comment (jrose-apple) [reply lib/FrontendTool/FrontendTool.cpp]: I see that it was already being used in the happy path and we should probably fix that in a follow-up.

File: lib/FrontendTool/FrontendTool.cpp [hunk 12]
```diff
@@ -237,12 +237,156 @@ template <> struct ObjectTraits<LoadedModuleTraceFormat> {
 }
 }
 
+namespace {
```
Comment (jrose-apple) [inline lib/FrontendTool/FrontendTool.cpp]: Instead of the anonymous namespace, we typically use `static` for helper functions. See the [LLVM Coding Standards](https://www.llvm.org/docs/CodingStandards.html#anonymous-namespaces).

File: lib/FrontendTool/FrontendTool.cpp [hunk 13]
```diff
@@ -237,12 +237,156 @@ template <> struct ObjectTraits<LoadedModuleTraceFormat> {
...
+        buffer.clear();
+      }
+
+      // FIXME: Better error handling
+      StringRef realDepPath
+        = fs::real_path(depPath, buffer, /*expand_tile*/true)
+        ? StringRef(depPath) // Couldn't find the canonical path, assume
+                             // this is good enough.
+        : buffer.str();
+      buffer.clear();
```
Comment (jrose-apple) [inline lib/FrontendTool/FrontendTool.cpp]: You can't clear the buffer; it's backing the StringRef.

Comment (varungandhi-apple) [reply lib/FrontendTool/FrontendTool.cpp]: Wow, I'm surprised to see that StringRef doesn't copy the data when you give it a raw pointer... How was this code even working before then? https://github.com/apple/swift/blob/a51162290e9df4507c1d5dcce58a0958dac4d785/lib/FrontendTool/FrontendTool.cpp#L284-L311

Comment (jrose-apple) [reply lib/FrontendTool/FrontendTool.cpp]: StringRef is string_view, the point is that it does not copy the data. I assume the old code worked because `real_path` is clearing the buffer contents itself, but it certainly doesn't say that in its doc comment. :-/

File: lib/FrontendTool/FrontendTool.cpp (hunk 10 above)
Comment (varungandhi-apple) [reply lib/FrontendTool/FrontendTool.cpp]: I think `find(..) != end()` is a fairly standard idiom, so I don't want to use `count != 0` here.

File: lib/FrontendTool/FrontendTool.cpp (hunk 11 above)
Comment (varungandhi-apple) [reply lib/FrontendTool/FrontendTool.cpp]: `exists` doesn't document what expansions it does, if any.

File: lib/FrontendTool/FrontendTool.cpp (hunk 10 above)
Comment (jrose-apple) [reply lib/FrontendTool/FrontendTool.cpp]: `count` on sets is [just as standard](https://en.cppreference.com/w/cpp/container/set/count). If you don't want to use that, though, you could go with `llvm::is_contained`.

File: lib/FrontendTool/FrontendTool.cpp (hunk 13 above)
Comment (varungandhi-apple) [reply lib/FrontendTool/FrontendTool.cpp]: If `real_path` were clearing the buffer, then we would've printed out the same path for every module, which wasn't happening earlier, because all the `realDepPath`s saved in the `swiftModules` vector point to the beginning of `buffer`.

File: lib/FrontendTool/FrontendTool.cpp (hunk 11 above)
Comment (jrose-apple) [reply lib/FrontendTool/FrontendTool.cpp]: Right, but you're not looking for expansions here. You're looking for an actual file next to where the swiftmodule was found.

File: lib/FrontendTool/FrontendTool.cpp (hunk 13 above)
Comment (jrose-apple) [reply lib/FrontendTool/FrontendTool.cpp]: SwiftModuleTraceInfo stores a std::string for the path; that's where it gets copied out.

File: lib/FrontendTool/FrontendTool.cpp (hunk 11 above)
Comment (varungandhi-apple) [reply lib/FrontendTool/FrontendTool.cpp]: I don't understand what you mean. With `exists`, if I give it a path `~/path/to/Foo.swiftinterface`, then I don't know if it will expand the `~` or not before checking if the file exists or not.

Put another way, it is undocumented what `exists` does for non-absolute paths. `moduleAdjacentInterfacePath` may not be an absolute path, because `depPath` may not be an absolute path.

File: lib/FrontendTool/FrontendTool.cpp (hunk 10 above)
Comment (varungandhi-apple) [reply lib/FrontendTool/FrontendTool.cpp]: > count on sets is just as standard.

What I meant by "standard idiom" is that `find() != end()` is a commonly understood idiom for `contains()`. Nothing complex about it. Replacing it with `pathToModuleDecl.count(..) != 0` feels more roundabout because we're not trying to _count_, we're trying to do a _find/lookup_.

Moreover, I have a map, not a set. This means that I cannot do `llvm::is_contained` because that needs a `DenseMapPair<StringRef, ModuleDecl *>`.

File: lib/FrontendTool/FrontendTool.cpp (hunk 8 above)
Comment (varungandhi-apple) [reply lib/FrontendTool/FrontendTool.cpp]: I will make this change in a cleanup PR later.

File: lib/FrontendTool/FrontendTool.cpp (hunk 11 above)
Comment (jrose-apple) [reply lib/FrontendTool/FrontendTool.cpp]: You know that `depPath` is a path that got looked up in some way, and you're forming a path relative to that. You'd need to be consistent with however the dependency was found, and you don't know what logic that was.  Expanding tildes would probably be wrong, even.

File: lib/FrontendTool/FrontendTool.cpp (hunk 4 above)
Comment (varungandhi-apple) [reply lib/FrontendTool/FrontendTool.cpp]: I've added a line of context on top.

File: lib/FrontendTool/FrontendTool.cpp (hunk 11 above)
Comment (varungandhi-apple) [reply lib/FrontendTool/FrontendTool.cpp]: (Punted to later.)

**Review by jrose-apple — APPROVED** [review]

**Comment by varungandhi-apple** [comment]
@swift-ci please smoke test and merge

**Comment by varungandhi-apple** [comment]
Why didn't it work 😕?

@swift-ci please smoke test
